### PR TITLE
Handle open types to appear in interface maps

### LINKED ISSUE #97667: TypeLoadException on minimal, correct and compiling program
### Description

A small program which compiles under .NET 6-8 and executes correctly on .NET 6-7, crashes on .NET 8 with a TypeLoadException about a failed generic constraint. Manual inspection shows that this constraint is not violated. For us (my company), this is a serious bug, as it completely prevents us from moving to .NET 8.

### Reproduction Steps

Use `dotnet run` on a console program with the following code:

```csharp
var a = new LifSharedVersion<object>();

public interface ILifVersionReadable<TA> {}

public class LifVersion<TVersion, TIVersionReadable>
    where TVersion : TIVersionReadable
{}

public class LifSharedVersion<TSharedVersionData> :
    LifVersion<LifSharedVersion<TSharedVersionData>, ILifSharedVersionReadable<TSharedVersionData>>,
    ILifSharedVersionReadable<TSharedVersionData>
{}

public interface ILifSharedVersionReadable<TSharedVersionData> :
    ILifVersionReadable<LifSharedVersion<TSharedVersionData>>
{}
```

This is a reduction from the bug we found in our code base. It is not immediately possible to reduce the program further, without the error going away.

This crash also occurs when compiling for .NET 6 and using `DOTNET_ROLL_FORWARD=LatestMajor` or similar.

### Expected behavior

The program completes without output.

### Actual behavior

The program crashes with the following exception:

```
Unhandled exception. System.TypeLoadException: GenericArguments[0], 'LifSharedVersion`1[TSharedVersionData]', on 'LifVersion`2[TVersion,TIVersionReadable]' violates the constraint of type parameter 'TVersion'.
   at Program.<Main>$(String[] args)
```

that is, System.TypeLoadException: GenericArguments[0], 'LifSharedVersion\`1[TSharedVersionData]', on 'LifVersion\`2[TVersion,TIVersionReadable]' violates the constraint of type parameter 'TVersion'.

### Regression?

There is no crash on .NET 6 or .NET 7.

### Known Workarounds

Removing anything from the reproduction example removes the problem. However, this reproduction example is a massively reduced version of a very large code base, so we have no easy way to work around this problem in what is supposed to soon be production code.

### Configuration

.NET 8, latest version. Otherwise, nothing special.

### Other information

_No response_

### PR DESCRIPTION
**Pull request by davidwrighton**
Fix the following scenarios
- Reflection IsAssignableFrom api
- As well as constraint checking

Fixes #97667

### CODE REVIEW

File: src/tests/Loader/classloader/generics/Instantiation/Positive/CuriouslyRecurringThroughInterface.cs [hunk 1]
```diff
@@ -13,15 +14,69 @@ class CuriouslyRecurringThroughInterface<T_CuriouslyRecurringThroughInterface> :
     {
     }
 
+    class BaseClassWhereAImplementsB<A,B> where A : B {}
+    interface ICuriouslyRecurring2<T_ICuriouslyRecurring> : IGeneric<DerivedCuriouslyRecurringThroughInterface<T_ICuriouslyRecurring>>
+    {
+    }
+    class DerivedCuriouslyRecurringThroughInterface<T_CuriouslyRecurringThroughInterface> : BaseClassWhereAImplementsB<DerivedCuriouslyRecurringThroughInterface<T_CuriouslyRecurringThroughInterface>,ICuriouslyRecurring2<T_CuriouslyRecurringThroughInterface>>, ICuriouslyRecurring2<T_CuriouslyRecurringThroughInterface>
+    {
+    }
```
Comment (lambdageek) [inline src/tests/Loader/classloader/generics/Instantiation/Positive/CuriouslyRecurringThroughInterface.cs:23]: So in this example, is it `BaseClassWhereAImplementsB` that gets the `MayHaveOpenInterfaceInInterfaceMap` bit set?

**Comment by lambdageek** [comment]
COM interop also has a call to `IsSpecialMarkerTypeForGenericCasting`

https://github.com/dotnet/runtime/blob/74597975f88355647e68add0623f2caf32d98c7f/src/coreclr/vm/classcompat.cpp#L1244-L1245

COM can't have the curiously recurring generic scenario, though, right?
Should we assert anyway?

File: src/tests/Loader/classloader/generics/Instantiation/Positive/CuriouslyRecurringThroughInterface.cs [hunk 2]
```diff
@@ -13,15 +14,69 @@ class CuriouslyRecurringThroughInterface<T_CuriouslyRecurringThroughInterface> :
...
+        {
+            Console.WriteLine("TestIfCuriouslyRecurringInterfaceCanBeUsedAsConstraint");
+            TestIfCuriouslyRecurringInterfaceCanBeUsedAsConstraintWorker();
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static void TestIfCuriouslyRecurringInterfaceCanBeUsedAsConstraintWorker()
+        {
+            // Test that the a generic using a variant of the curiously recurring pattern involving an interface and constraint can be loaded.
+            // This test is just like TestIfCuriouslyRecurringInterfaceCanBeLoaded, except that it is structured so that we perform a cast via a constraint at type load time
```
[the remark this answers is not in the document]
Comment (davidwrighton) [reply src/tests/Loader/classloader/generics/Instantiation/Positive/CuriouslyRecurringThroughInterface.cs:77]: These don't need to run isolated as we have separate types for the curiously recurring pattern for the two tests, so the type loading operations will be independent. However, yes, we do have a scheme for running the tests in separate processes. First you need to have them be in separate project files, then you put `<RequiresProcessIsolation>true</RequiresProcessIsolation>` into the project files.

File: src/tests/Loader/classloader/generics/Instantiation/Positive/CuriouslyRecurringThroughInterface.cs (hunk 1 above)
Comment (davidwrighton) [reply src/tests/Loader/classloader/generics/Instantiation/Positive/CuriouslyRecurringThroughInterface.cs:23]: No actually its `DerivedCuriouslyRecurringThroughInterface<T_ICuriouslyRecurring>` which has the `MayHaveOpenInterfaceInInterfaceMap` bit set.

File: src/tests/Loader/classloader/generics/Instantiation/Positive/CuriouslyRecurringThroughInterface.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (lambdageek) [reply src/tests/Loader/classloader/generics/Instantiation/Positive/CuriouslyRecurringThroughInterface.cs:77]: Oh, I missed that they were working on separate types.

**Comment by davidwrighton** [comment]
> COM interop also has a call to `IsSpecialMarkerTypeForGenericCasting`
> 
> https://github.com/dotnet/runtime/blob/74597975f88355647e68add0623f2caf32d98c7f/src/coreclr/vm/classcompat.cpp#L1244-L1245
> 
> COM can't have the curiously recurring generic scenario, though, right? Should we assert anyway?

Unfortunately, we can't effectively assert there. We might or might not have the flag set, and in either case it doesn't really make any sense in the presence of COM interop.

**Comment by lambdageek** [comment]
@davidwrighton Can we do a call Thursday or Friday to review. I'm having some trouble following what's going on (new codebase + curiously recursive generics == :exploding_head: )

**Comment by davidwrighton** [comment]
@lambdageek, absolutely, just schedule drop something onto my schedule

**Review by lambdageek — APPROVED** [review]
I also can't make the example convoluted enough to hit the variant case without making the bug go away (because I end up passing at least one generic type argument that is not just a generic parameter - so we don't get the typical instantiation).

The code as written matches what we thought it ought to do during the code review 😄

**Pull request merged**